components/url/lib.rs

```diff
@@ -25,6 +24,29 @@ use url::{Position, Url};
 
 pub use crate::origin::{ImmutableOrigin, MutableOrigin, OpaqueOrigin};
 
+#[derive(Debug)]
+pub enum UrlError {
+    SetUsername,
+    SetIpHost,
+    SetPassword,
+    ToFilePath,
+    FromFilePath,
+}
+
+impl std::fmt::Display for UrlError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
```

Comment:
A small nit, I think it would make sense if you used the already imported `fmt::` namespace for `Display`, `Formatter` and `Result`, as most of the files in the project already use this convention. If you decide to do it, it would also be nice to update the return type of `debug_compact` to match.

```suggestion
impl fmt::Display for UrlError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```

On a more global note, I wonder if it is necessary to add this error type. This is wrapping `url`'s library functions, and they decided not to include an error type for [these functions](https://docs.rs/url/latest/url/struct.Url.html#method.set_username). As I saw it the backtrace in case of an error would be enough to determine which function was causing the issue, that's why I added the `allow`, but a case could be made that the error provides better context.